pkg/buildpack/buildpack.go

```diff
@@ -326,6 +356,131 @@ func ToLayerTar(dest string, module BuildModule) (string, error) {
 	return layerTar, nil
 }
 
+func ToNLayerTar(dest string, module BuildModule, logger logging.Logger) ([]ModuleTar, error) {
```

Comment:
Maybe it's just me, but personally I found this function a bit hard to read and understand. The need to keep track of `rootFolderHeaders` and then `continue` within the loop was a bit confusing. What about something like:

```
func ToNLayerTar(dest string, module BuildModule) ([]ModuleTar, error) {
	if !module.ContainsFlattenedModules() {
		return handleSingleOrEmptyModule(dest, module)
	}

	modReader, err := module.Open()
	if err != nil {
		return nil, errors.Wrap(err, "opening blob")
	}
	defer modReader.Close()

	tarCollection := newModuleTarCollection(dest)
	tr := tar.NewReader(modReader)

	// read the first header
	header, err := tr.Next()
	if err != nil {
		if err == io.EOF {
			return handleSingleOrEmptyModule(dest, module)
		}
		return nil, fmt.Errorf("failed to read first header '%s': %w", header.Name, err)
	}

	// the original version should be blank because the first header should look like "/cnb/buildpacks/<buildpack-id>"
	origID, origVersion := parseBpIDAndVersion(header)
	if origVersion != "" {
		return nil, fmt.Errorf("first header '%s' contained unexpected version", header.Name)
	}

	if err := toNLayerTar(origID, origVersion, header, tr, tarCollection); err != nil {
		return nil, err
	}

	errs := tarCollection.close()
	if len(errs) > 0 {
		return nil, errors.New("closing files")
	}

	return tarCollection.getModules(), nil
}
```

That uses a recursive function:
```
func toNLayerTar(origID, origVersion string, firstHeader *tar.Header, tr *tar.Reader, tc *moduleTarCollection) error {
	toWrite := []*tar.Header{firstHeader}
	if origVersion == "" { // TODO: test flattened module that contains buildpacks with same ID but different versions
		// the first header only contains the id - e.g., /cnb/buildpacks/<buildpack-id>,
		// read the next header to get the version
		secondHeader, err := tr.Next()
		if err != nil {
			return err
		}
		nextID, nextVersion := parseBpIDAndVersion(secondHeader)
		if nextID != origID || nextVersion == "" {
			return fmt.Errorf("second header '%s' contained unexpected id or missing version", secondHeader.Name)
		}
		origVersion = nextVersion
		toWrite = append(toWrite, secondHeader)
	} else {
		// the first header contains id and version - e.g., /cnb/buildpacks/<buildpack-id>/<buildpack-version>,
		// we need to write the parent header - e.g., /cnb/buildpacks/<buildpack-id>
		realFirstHeader := firstHeader
		realFirstHeader.Name = filepath.Dir(firstHeader.Name) // TODO: check this logic
		toWrite = append([]*tar.Header{realFirstHeader}, toWrite...)
	}
	mt, err := tc.get(origID, origVersion)
	if err != nil {
		return err
	}
	for _, h := range toWrite {
		if err := mt.writer.WriteHeader(h); err != nil {
			return fmt.Errorf("failed to write header '%s': %w", h.Name, err)
		}
	}
	// write the rest of the package
	var header *tar.Header
	for {
		header, err = tr.Next()
		if err != nil {
			if err == io.EOF {
				return nil
			}
			return err
		}
		nextID, nextVersion := parseBpIDAndVersion(header)
		if nextID != origID || nextVersion != origVersion {
			// we found a new module, recurse
			return toNLayerTar(nextID, nextVersion, header, tr, tc)
		}

		err = mt.writer.WriteHeader(header)
		if err != nil {
			return errors.Wrapf(err, "failed to write header for '%s'", header.Name)
		}

		buf, err := io.ReadAll(tr)
		if err != nil {
			return errors.Wrapf(err, "failed to read contents of '%s'", header.Name)
		}

		_, err = mt.writer.Write(buf)
		if err != nil {
			return errors.Wrapf(err, "failed to write contents to '%s'", header.Name)
		}
	}
}
```

Reply:
Well, the code works fine, all the tests and acceptance tests are passing.

Reply:
This code ^^ seems clearer to me, but I wrote it 🙃 @joe-kimmel-vmw do you have any opinions about which one is friendlier to a newcomer?

Reply:
Adding the test case you mentioned in the `TODO: test flattened module that contains buildpacks with same ID but different versions
the first header only contains the id - e.g., /cnb/buildpacks/<buildpack-id>,
read the next header to get the version`


I am getting the error:

```bash
buildpack_test.go:776: Expected nil: could not find entry path '/cnb/buildpacks/buildpack-1-id/buildpack-1-version-2': not exist
```

But I am also having an error with the previous code, so I will take a look